startup/81-pe-callbacks.py

```diff
@@ -68,11 +108,123 @@ def event_page(self, doc):
+
+    @property
+    def configuration_attrs(self):
+        return self._configuration_attrs
+
+    @property
+    def read_attrs(self):
+        return self._read_attrs
+
+    def describe(self):
+        return self._describe
+
+    def describe_configuration(self):
+        return self._describe_configuration
+
+    # def describe_configuration(self):
+    #     return self.det.describe_configuration
+
+    def collect_asset_docs(self):
+        # keep track of when we get restaged to restore these
+        yield from self._asset_docs_cache
+        self._really_cached = self._asset_docs_cache
+        self._asset_docs_cache = []
+
+    def stage(self):
+        self._asset_docs_cache = self._really_cached
+
+
+def dark_plan(cam, dark_frame_cache, obsolete_secs, shutter):
+    if (dark_frame_cache.just_started or  # first run after instantiation
+        (dark_frame_cache.last_collected is not None and
+         time.monotonic() - dark_frame_cache.last_collected > obsolete_secs)):
+        init_shutter_state = shutter.status.get()
+        yield from bps.mv(shutter, 'Close')
+        yield from bps.trigger(cam, group='cam')
+        yield from bps.wait('cam')
+        yield from bps.mv(shutter, init_shutter_state)
+
+        teleport(cam, dark_frame_cache)
+        dark_frame_cache.just_started = False
+        dark_frame_cache.update_done = True
+    else:
+        dark_frame_cache.update_done = False
+
+
+def teleport(cam, dfc):
+    dfc._describe = cam.describe()
+    dfc._describe_configuration = cam.describe_configuration()
+    dfc._read = cam.read()
+    dfc._read_configuration = cam.read_configuration()
+    dfc._read_attrs = cam.read_attrs
+    dfc._configuration_attrs = cam.configuration_attrs
+    dfc._asset_docs_cache = list(cam.collect_asset_docs())
+    dfc.last_collected = time.monotonic()
+
+
+dark_frame_cache = DarkFrameCache(name='dark_frame_cache')
+# Update the '_<name>' attributes which do not exist yet
+teleport(pe1c, dark_frame_cache)  # noqa
+
+
+def dark_frame_aware_plan(cam, dark_frame_cache, shutter=shutter_fs,
+                          obsolete_secs=60, num_images=1, md=None):
+
+    def check_and_take_darks():
+        yield from dark_plan(cam, dark_frame_cache, obsolete_secs, shutter)
+        if dark_frame_cache.update_done:
+            yield from bpp.trigger_and_read([dark_frame_cache], name='dark')
+
+    @bpp.stage_decorator([cam])
+    @bpp.run_decorator(md=md)
+    def inner_dark_frame_aware_plan():
+        init_shutter_state = shutter.status.get()
+        yield from bps.mv(shutter, 'Open')
+
+        for _ in range(num_images):
+            yield from check_and_take_darks()
+            yield from bpp.trigger_and_read([cam], name='primary')
+
+        yield from bps.mv(shutter, init_shutter_state)
+
+    return (yield from inner_dark_frame_aware_plan())
+
+
+def subtract_dark(light_img, dark_img):
+    res = np.asarray(light_img, dtype=int) - np.asarray(dark_img, dtype=int)
+    res[res < 0] = 0
```

Comment:
you can use `np.clip` here with I suspect will be more efficient.

Reply:
OK, that sounds good.